jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -1794,29 +1794,33 @@ public boolean equalsIgnoreCase(String string) {
 			byte byteAtO1Last = helpers.getByteFromArrayByIndex(s1Value, s1len - 1);
 			byte byteAtO2Last = helpers.getByteFromArrayByIndex(s2Value, s1len - 1);
 
-			if (byteAtO1Last != byteAtO2Last &&
-					toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last))) {
+			if ((byteAtO1Last != byteAtO2Last)
+					&& (Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO1Last)) != Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO2Last)))
+			) {
 				return false;
 			}
 
 			while (o1 < end - 1) {
 				byte byteAtO1 = helpers.getByteFromArrayByIndex(s1Value, o1++);
 				byte byteAtO2 = helpers.getByteFromArrayByIndex(s2Value, o2++);
 
-				if (byteAtO1 != byteAtO2 &&
-						toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != toUpperCase(helpers.byteToCharUnsigned(byteAtO2))) {
+				if ((byteAtO1 != byteAtO2)
+						&& (Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO1)) != Character.toUpperCase(helpers.byteToCharUnsigned(byteAtO2)))
+				) {
 					return false;
 				}
 			}
 		} else {
-			// Compare the last chars. Under string compression, the compressible char set obeys 1-1 mapping for upper/lower
-			// case, converting to lower cases then compare should be sufficient.
+			// Compare the last chars.
+			// In order to tell 2 chars are different:
+			// If at least one char is ASCII, converting to upper cases then compare should be sufficient.
+			// If both chars are not in ASCII char set, need to convert to lower case and compare as well.
 			char charAtO1Last = s1.charAtInternal(s1len - 1, s1Value);
 			char charAtO2Last = s2.charAtInternal(s1len - 1, s2Value);
 
-			if (charAtO1Last != charAtO2Last
-					&& toUpperCase(charAtO1Last) != toUpperCase(charAtO2Last)
-					&& ((charAtO1Last <= 255 && charAtO2Last <= 255) || Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))
+			if ((charAtO1Last != charAtO2Last)
+					&& (Character.toLowerCase(Character.toUpperCase(charAtO1Last)) != Character.toLowerCase(Character.toUpperCase(charAtO2Last)))
+					&& (charAtO1Last <= 255 && charAtO2Last <= 255)
```

Comment:
Existing code checks `(charAtO1Last <= 255 && charAtO2Last <= 255)` along with `|| Character.toLowerCase(charAtO1Last) != Character.toLowerCase(charAtO2Last))`, now it is required alone as one of conditions, it doesn't seem right.
Does this change pass a personal build test suite like `sanity.functional`?

Reply:
Currently waiting on the results of my personal build test suite with the current pushed changes, will update once available.

In the meantime, I tested locally using `(char1 != char2) && (compareValue(char1) != compareValue(char2)` (removed `(charAtO1Last <= 255 && charAtO2Last <= 255)` and swapped `Character.toLowerCase(Character.toUpperCase()` with `compareValue()`), and the expected behaviour was seen. This logic is currently being used here on line `1839`:
https://github.com/eclipse/openj9/blob/f2e8dfa4346808869c9e47822f725a10678ac97d/jcl/src/java.base/share/classes/java/lang/String.java#L1836-L1841

`compareValue(char/int/byte)` internally calls `Character.toLowerCase(Character.toUpperCase(char/int/byte))`

https://github.com/eclipse/openj9/blob/f2e8dfa4346808869c9e47822f725a10678ac97d/jcl/src/java.base/share/classes/java/lang/String.java#L1457-L1478

I'll run another build with these changes to confirm if it's functionally correct.

Reply:
Both the build with the currently pushed code and the build with my local changes passed `sanity.functional`. Will be pushing my local changes, which contain the modifications discussed in the review comments.